File: NutriNotion-backend/src/main/java/com/nutrinotion/nutrinotion_backend/model/PersonalizedMealItem.java
package com.nutrinotion.nutrinotion_backend.model;

import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDate;

/**
 * One food item inside a daily personalized meal plan.
 * Rows are created when the AI generates today's plan.
 * isChecked is toggled by the user as they eat each item.
 */
@Entity
@Data
@Table(name = "personalized_meal_items")
public class PersonalizedMealItem {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "meal_date", nullable = false)
    private LocalDate mealDate;

    /**
     * "breakfast" | "lunch" | "snacks" | "dinner"
     */
    @Column(name = "meal_type", nullable = false, length = 20)
    private String mealType;

    @Column(name = "food_name", nullable = false, length = 500)
    private String foodName;

    /**
     * Human-readable quantity — e.g. "1 serving", "200g", "1 cup"
     */
    @Column(length = 100)
    private String quantity;

    /**
     * Estimated calories for this item/quantity.
     */
    private Integer calories;

    /**
     * Toggled by PUT /api/personalized-meals/item/{id}/check.
     * When set to true → calories are added to DailyCalorieTracking.
     * When set to false → calories are subtracted.
     */
    @Column(name = "is_checked", nullable = false)
    private boolean isChecked = false;
}

File: NutriNotion-Frontend/lib/models/personalized_meal_item.dart
/// Represents a single food item in a personalized meal plan.
class PersonalizedMealItem {
  final int? id;
  final String foodName;
  final String quantity;
  final int calories;
  final String mealType;
  bool isChecked;

  PersonalizedMealItem({
    this.id,
    required this.foodName,
    required this.quantity,
    required this.calories,
    required this.mealType,
    this.isChecked = false,
  });

  factory PersonalizedMealItem.fromJson(Map<String, dynamic> json) {
    return PersonalizedMealItem(
      id: json['id'] as int?,
      foodName: (json['foodName'] ?? json['item'] ?? '').toString(),
      quantity: (json['quantity'] ?? '').toString(),
      calories: _parseCalories(json['calories']),
      mealType: (json['mealType'] ?? '').toString(),
      isChecked: json['isChecked'] == true,
    );
  }

  static int _parseCalories(dynamic v) {
    if (v is int) return v;
    if (v is num) return v.toInt();
    if (v is String) return int.tryParse(v) ?? 0;
    return 0;
  }

  PersonalizedMealItem copyWith({bool? isChecked}) {
    return PersonalizedMealItem(
      id: id,
      foodName: foodName,
      quantity: quantity,
      calories: calories,
      mealType: mealType,
      isChecked: isChecked ?? this.isChecked,
    );
  }

  @override
  String toString() =>
      'PersonalizedMealItem(id=$id, foodName=$foodName, qty=$quantity, cal=$calories, mealType=$mealType, checked=$isChecked)';
}
